Implement #8020: Add age of password to health check

mattesony opened the pull request:
Implements #8020. I have only added in a check for older than 1 year and a starter penalty for now; I would be willing to expand if/when I have more time.

The only change to existing code is making changes to the entry health scoring; everything else is additive. It takes 5 points away for every year over 1 in age and notates the reason. I didn't put much though into either number -- it doesn't seem to guarantee that old passwords show up if only 5 points are taken off. I suggested a few potential alternate approaches in #8020.

## Screenshots
![image](https://user-images.githubusercontent.com/49170923/199212183-88edfc76-3d67-404a-abe7-a0569925585a.png)

## Testing strategy
Manual testing, unit tests to come if this approach looks good to those more knowledgeable than I.

## Type of change
- ✅ New feature (change that adds functionality)

mattesony committed:
First pass at adding age penalty to health check

droidmonkey commented:
Don't reduce the score, just post the text. I like your password age function though, that will be useful in other locations.

mattesony commented:
> Don't reduce the score, just post the text. I like your password age function though, that will be useful in other locations.

I'm not very attached to the scoring, so I'm fine with dropping it, but how does someone see passwords that are old if there's nothing else to report about them, since they won't show up?

mattesony committed:
Removed penalty for old age

droidmonkey commented:
We can introduce a "flagged" boolean for the rating so it shows. Can also set it to show if there are any score reasons.

mattesony committed:
Switched to seconds for getPasswordAge for better reuse

mattesony commented:
Since it seemed like the age function would be useful in other places, I changed it to seconds to stay in line with the age calcs for history entries. It also let me use `Tools::humanReadableTimeDifference(seconds)` to keep the descriptions of the ages consistent with those in the history viewer:
![image](https://user-images.githubusercontent.com/49170923/199228530-d0230a6d-fdc4-4ef5-926a-88013c609922.png)

mattesony committed:
Fixed outdated comments

mattesony committed:
Removed one other extraneous line

mattesony committed:
getPasswordAge unit tests.
Tests fail because QDateTime::currentDateTime() is not using the MockClock instance.
The time is correct, but as it changes every time it makes the test non-deterministic.

mattesony committed:
Fixed tests and improved age logic

mattesony committed:
Formatting

mattesony commented:
I added unit tests for the `getPasswordAge()` function. I couldn't find tests for the score reasons; I'm not sure if those are necessary for this change alone and I'm not hugely confident with the tests (yet), so I'm not adding any.

mattesony committed:
Translations

droidmonkey commented on src/core/Entry.h:
```diff
@@ -116,6 +116,7 @@ class Entry : public ModifiableObject
     QString path() const;
     const QSharedPointer<PasswordHealth> passwordHealth();
     const QSharedPointer<PasswordHealth> passwordHealth() const;
+    int getPasswordAge() const;
```
Note to self, rename to `passwordAgeSeconds()`

mattesony replied:
I renamed it.

droidmonkey commented on tests/TestEntry.cpp:
```diff
@@ -47,6 +49,58 @@ void TestEntry::testHistoryItemDeletion()
     QVERIFY(historyEntry.isNull());
 }
 
+void TestEntry::testGetPasswordAge()
```
Rename `testPasswordAge`

mattesony replied:
I renamed this too, but I'm just realizing that I renamed it to `testPasswordAgeSeconds` -- do you still want `testPasswordAge`? I'm happy to change it, if so.

droidmonkey replied:
That name is fine

droidmonkey commented on tests/TestEntry.cpp:
```diff
@@ -47,6 +49,58 @@ void TestEntry::testHistoryItemDeletion()
 
+void TestEntry::testGetPasswordAge()
+{
+    MockClock* m_clock = new MockClock(2022, 2, 4, 17, 00, 00);
```
Rename to `mockClock`

mattesony replied:
Done!

mattesony committed:
Rename Entry::getPasswordAge() to Entry::passwordAgeSeconds()

mattesony committed:
Merge branch 'develop' into 8020-age

pamperer562580892423 commented:
Hey, I'm just an interested lay person and saw this here. Instead of 'score' it would be great to see the entropy of the password. Even it is marked as 'excellent' or 'strong', the entropy gives a lot more info and one could decide to set up new passwords with even higher entropy, if one wants to even get better "health" of one's passwords.

droidmonkey commented:
Entropy only applies to the password. The score uses more than just the password Entropy to come to a value. The exact Entropy of a password is relatively meaningless, and our poor, weak, good, excellent ratings are calibrated to certain levels of Entropy.

pamperer562580892423 commented:
@droidmonkey Thanks for the reply. At first, I didn't want to reply to that because my uninformed opinion is irrelevant. But on the other hand: up until now I thought that entropy is the most important thing for a password - and commonly used. And my thought was: "excellent" begins with an entropy of 100 at the moment. But most of the time I go up even higher when I can, e.g. at least 128 or 200 bits of entropy. And this differentiation would be really nice to see in this overview.
And I think as you stated in another "pull" (https://github.com/keepassxreboot/keepassxc/issues/8020) the age of a password doesn't really reduce it's security, so to see a "score" (as a compound value) instead of (solely) entropy, would make things more complicated for me.
Apart from that: maybe an optional rule like "show/report me all passwords, which are older than XYZ days/years" would be interesting, because every few years, passwords may be reasonably adapted to changing computational/hacking power etc.?

droidmonkey commented:
Passwords do not "age", but they do get more likely to have been compromised the longer they are used. That is just simple statistics, not security. Entropy is a measure of randomness (really energy) and a password with higher entropy are harder to brute force (ie guess). Entropy is just a guess of randomness since there are a lot of factors that go into it, which is why you should be less concerned with the _exact value_ and more concerned with the broad bin that value places you in (poor, weak, excellent, etc). 

Over time we will continue to increase the entropy required to reach the excellent bin, we have already increased it once.

pamperer562580892423 commented:
I'm not concerned about the exact value of entropy - but as the example: 200 instead of 100 makes a difference, though both "excellent". And yes, statistics is a good point.
Anyway, as a further "user feedback": I just searched for an info about the score in the user guide and either I'm blind or stupid or there is no info about what the score consists of. It would be interesting to know, what builds up the score so that the user can understand it better. For me it is too abstract and of little value because of that. (and as I said, entropy in combination with info about the age of the password - as separate infos - would be more valuable and less "abstract". (by the way: I don't see me as totally uninformed, so I think many less versed users would also be more able to grasp "entropy" and "age" separately much better than a "compound value", especially when the "score" is not explained)

pamperer562580892423 commented:
(BTW: I hope this is seen as a "constructive feedback" - wheter relevant or irrelevant - I don't want to just "whinge". I'm really thankful of your work and appreciate this password manager and it's richness in functions very much.)

droidmonkey commented:
> there is no info about what the score consists of

The reports view tells you what is affecting the score in the far right column. Score is just a way to sort entries by "worst to best".

michaelk83 commented:
There are diminishing returns as you go higher in entropy. If it takes 2 billion years to brute-force your password, or 1 billion years, it doesn't really matter at that point. So once you hit the "excellent" bin, any further entropy doesn't really add to security in a practical sense.

As for distinction of the factors that go into the score - all you really need to care about is just updating your password once it goes below "good" (or below "excellent", if you prefer). That would automatically take care of both the age and the entropy.  Use the password generator to get a good enough score, and you should be good to go (the dice icon inside the password field).
